source/ImageBuffer.cpp

```diff
@@ -300,6 +298,19 @@ namespace {
 		jpeg_finish_decompress(&cinfo);
 		jpeg_destroy_decompress(&cinfo);
 		
+		// Expand RGB to RGBA
+		for (unsigned int y = 0; y < cinfo.image_height; y++)
+		{
+			auto pptr = reinterpret_cast<uint8_t*>(buffer.Begin(y, frame));
```

Comment:
When going from RGB (three 8-bit values) to RGBA (four 8-bit values), then I would expect different starting points for all lines after the first line. (And I would expect the for-loop above to loop from height back to 0 to avoid the risk of overwriting existing data).
Also, isn't there some library function available in libJPEG to handle this conversion (since I can imagine that different platforms actually store things differently)?

Reply:
it already does strides in reverse order - see the loop below.

What do you mean by "different starting points for all lines after the first line"? I request the pointer to a specific line in the buffer already, but I feel you meant something else?

> some library function available in libJPEG to handle this conversion

Honestly haven't thought of that, but I can check - my reasoning was that if such a format was supported, user would be able to request it explicitly when decoding (like libjpeg-turbo does) to avoid duplicate work, so it makes little sense to add such a trivial function to the library.

Reply:
Haven't seen anything interesting in https://github.com/emscripten-ports/SDL2_image/blob/master/external/jpeg-9/jpeglib.h

Reply:
I was surprised that the strides all start at the same location. (Both the source and target strides seem to start from the `auto pptr`.) This suggests that the RGB image is having much larger strides than it would normally need.

Reply:
> > some library function available in libJPEG to handle this conversion
> 
> if such a format was supported, user would be able to request it explicitly when decoding (like libjpeg-turbo does) to avoid duplicate work, so it makes little sense to add such a trivial function to the library.

Such a function would be trivial, until we get to a point where we need to compensate for differences in platforms (big-endian vs little-endian, 8-bit aligned vs 32-bit aligned vs 64-bit aligned memory layouts). The least that needs to be done when writing those functions ourselves is to test them for all supported platforms (to verify that we didn't miss anything platform specific).

Reply:
> I was surprised that the strides all start at the same location. (Both the source and target strides seem to start from the auto pptr.) This suggests that the RGB image is having much larger strides than it would normally need.

@petervdmeer, that's right, the code above here uses `cinfo.out_color_space = JCS_RGB;` so 24-bit pixels, but in the ImageBuffer class the space for image data is always 4 bytes per pixel:
https://github.com/endless-sky/endless-sky/blob/ffd6053b9c475a01ed48b4758489ca17f079e3f2/source/ImageBuffer.cpp#L68

so `buffer.Begin(y, frame))` will be the same for both encodings, but the location of each pixel in the line will be different. Would you suggest changing this, maybe by using an out_color_space of JCS_RGBX if that exists? Or maybe just a comment describing this?

Regarding tests, I could use a pointer on how to write tests for this; is it just a new unit test that decodes an image and tests whether a couple pixels match what we expect?

Reply:
A comment to describe why the start positions are the same would be sufficient for me.

A unit test that decodes an image and checks a couple of pixels is indeed the test-size I was thinking about. The nice thing of such unit-test is that we can automatically run it on all platforms where ES is supported (even on the web-assembly version if you like).
Here is a nice [starting point for an unit testing framework](https://github.com/endless-sky/endless-sky/blob/master/tests/src/README.md) that is already in use for ES.

Reply:
I'm looking into this now, it seems to me I'll need to change the way the tests are run to link in libjpeg.

Reply:
You won't be able to use third party code such as libpng or jpg-turbo from the tests without linking in their code, which is intentionally not done for a unit test. Instead one should provide a mock with the appropriate endpoints to be consumed by the class, which could then be inspected to ensure the code under test used the library correctly.

I haven't delved deeply into how exactly one would provide a mock library implementation in the tests that would be used by the game's class with our setup, but I imagine it would not be trivial. Isolating each test's code from other tests naturally means you also isolate it from the game classes' code, so you'd need to write an actual shared library and link it, which means updating scons and codeblocks projects so that it works on CI and also for windows devs using CB instead of scons.

I'm totally open to implementing our unit tests in some other framework or manner that makes it easier to test, but a core requirement I have is that the tests have to be manageable and runnable from CodeBlocks.

Reply:
Since it sounds like this will be some work, @petervdmeer could I convince you that endianness won't be a problem because these are all single-byte values and that regardless of alignment, RGB output colorspace will always be 24 bits? I'm not sure about that second one but reading around it seems to be true.

Reply:
As I said before: A number of good comments in the code on why this works is sufficient for me. The tests would be a nice extra, but not required.